### ISSUE TITLE
tab chars

### ISSUE THREAD

**Issue by AlexanderARodin** [preiter93/edtui#14]
there are a few problems with tab-chars:
1) loading content into the State with tab-chars inside leads to broken screen
2) on the other hand, the Handler expectantly ignores TAB key event

If I'm correct, tab-chars is not in the road-map. Or maybe it does work but I should dive deeper in the docs.
Anyway, it's not a difficult to do work-around but it's just not obvious enough.

**Comment by preiter93** [comment]
Makes sense to support. I will put it in the roadmap. Not sure how difficult its gonna be.

I simply havent noticed cause I rarely use tabs.

**Comment by preiter93** [comment]
Related to
https://github.com/ratatui/ratatui/issues/876#issue-2103844309

**Comment by AlexanderARodin** [comment]
seems to be no universal way to process tab-chars.
it needs to be done explicitly from case to case.
thanx for attention 👍

**Comment by preiter93** [comment]
I think we can still do it within this library. We can map a tab character to a configurable amount of spaces for the view. Somewhere in these 2 places
https://github.com/preiter93/edtui/blob/da8f076a1afa5adc6bfc58684f6d66738901f7f0/src/internal.rs#L200
https://github.com/preiter93/edtui/blob/da8f076a1afa5adc6bfc58684f6d66738901f7f0/src/internal.rs#L237
I will check on the weekend

I will reopen the isse until then as a reminder :)

**Comment by preiter93** [comment]
@AlexanderARodin I've added support for tabs, the code is on the "tabs" branch. Feel free to give it a try
```
edtui = { version = "0.9", git = "https://github.com/preiter93/edtui", branch = "tabs" }
```

You can set the tab width
```
EditorView::new(&mut state).tab_width(2);
```

I have to do some more testing before I merge this into main. 
It required some tweaking to get rendering of characters unicode widths != 1 right, which isnt easy. So if you find anything thats broken let me know!

**Comment by AlexanderARodin** [comment]
it works definitely better than my hot-fix-work-around. for example, it jumps on tabs in text.
as for me, it would be more cool, if it did vertical alignment on column (see the attachment).
![image](https://github.com/user-attachments/assets/f8a36010-3b26-420d-96ef-69f4554d33ab)
it would be more similar to way editors work with tabs.

**Comment by AlexanderARodin** [comment]
here is an example from nvim:
![image](https://github.com/user-attachments/assets/659fe8a2-b69a-4dd7-abb2-02a6487a40bd)

**Comment by AlexanderARodin** [comment]
moreover, looks like there is some issue with system buffer.
I tried to copy-past from EdTui to NeoVim and it was somehow mutated (the attachment)
![image](https://github.com/user-attachments/assets/db1cb408-a5fc-4081-9379-8229422f1b80)

**Comment by preiter93** [comment]
thx for checking, I appreciate it :)

I didn't realize the editors were doing something smart with tabs. The vertical alignment is something that would add a lot of complexity to the code without huge benefit, so I would give this a low priority for now unless theres a big motivation. But I do realize that most editors and terminals do the vertical alignment, so it would be nice to have in the future.

For the copying, how do you copy? If I visually select and copy with `p` in edtui and `y` in neovim I get the result as expected. I dont see how you got the 2tabs in edtui -> 4 tabs in nvim. The tabs `\t` stay untouched in the raw data and are only interpreted differently for rendering. But copying copies the raw data, so 2 tabs should just be copied as `\t\t`.

**Comment by AlexanderARodin** [comment]
understood. actually, I think, better way if tab-support was placed in Paragraph itself rather than some tricky way around it.

what about copying: now I'm working mostly under WSL, so in EdTui I used <y> for copy and then inserted in another WSL-session via window menu
![image](https://github.com/user-attachments/assets/22c95952-fb62-4c56-afe5-4f102e520597)

btw, I noticed that NeoVim itself can't use windows system clippboard but EdTui can xD

**Comment by AlexanderARodin** [comment]
And finally: as for me, that minimal support of Tabs is more beyond I hoped to. So thank you anyway. It's already better than those things I did around tabs.

**Comment by preiter93** [comment]
I will close this issue in favor of
https://github.com/preiter93/edtui/issues/18

**Comment by preiter93** [comment]
Published the tab support in 0.9.1
